extensions/PhabBugz/lib/Util.pm

```diff
@@ -0,0 +1,310 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This Source Code Form is "Incompatible With Secondary Licenses", as
+# defined by the Mozilla Public License, v. 2.0.
+
+package Bugzilla::Extension::PhabBugz::Util;
+
+use 5.10.1;
+use strict;
+use warnings;
+
+use Bugzilla::Error;
+
+use Bugzilla::Extension::PhabBugz::Constants;
+
+use Data::Dumper;
+use JSON qw(encode_json decode_json);
+use LWP::UserAgent;
+
+use base qw(Exporter);
+
+our @EXPORT = qw(
+  create_revision_attachment
+  create_private_revision_policy
+  create_project
+  edit_revision_policy
+  get_bug_role_phids
+  get_members_by_bmo_id
+  get_project_phid
+  get_revision_by_id
+  intersect
+  make_revision_public
+  request
+  set_project_members
+);
+
+sub get_revision_by_id {
+    my $id = shift;
+
+    my $data = {
+        queryKey    => 'all',
+        constraints => {
+            ids => [ int($id) ]
+        }
+    };
+
+    my $result = request( 'differential.revision.search', $data );
+
+    ThrowUserError('invalid_phabricator_revision_id')
+      unless ( exists $result->{result}{data} && @{ $result->{result}{data} } );
+
+    return $result->{result}{data}[0];
+}
+
+sub create_revision_attachment {
+    my ( $bug, $revision_id, $revision_title ) = @_;
+    my $user = Bugzilla->user;
+
+    my $phab_base_uri = Bugzilla->params->{phabricator_base_uri};
+    ThrowUserError('invalid_phabricator_uri') unless $phab_base_uri;
+
+    my $revision_uri = $phab_base_uri . "D" . $revision_id;
+
+    # Check for previous attachment and return if one matches exact
+    my @reviews =
+      grep { $_->contenttype eq PHAB_CONTENT_TYPE } @{ $bug->attachments };
+    foreach my $attachment (@reviews) {
+        return $attachment if $attachment->data eq $revision_uri;
+    }
+
+    # Automation user needs to be able to attach to private bugs
+    my $old_user     = Bugzilla->user;
+    my $current_user = Bugzilla->user;
+    $current_user->{groups} = [ Bugzilla::Group->get_all ];
+
```

Comment:
I think this is an error rust would catch!
$old_user and $current_user are references to the same object,
so changing the groups of current_user changes the old one too.

you'll need to do something more here... keep a copy of the original groups entry?